Implement a conversion plugin to bake noise textures

cfare opened the pull request:
### Overview
Implements a conversion plugin that allows NoiseTexture2D resources to be converted to WebP or PNG files and reimported as CompressedTexture2D.

The motivating use case for this is using a high-resolution noise texture that does not need to be altered after being initially configured - e.g. driving a wave effect in a shader. NoiseTexture2Ds are regenerated when the scene is loaded, and can take many seconds at high resolutions, which can be really annoying when trying to iterate quickly during development, or adversely impact load times in finished projects.

In my own use cases I rarely need to modify a noise texture after setting it up, and if that is required it's still simple enough to recreate it. I assume this is true of many users.

This is my first PR so please excuse me if I missed something obvious!

### Implementation
The existing conversion plugin design does not allow interactivity - conversions are performed immediately and cannot prompt the user for input. This PR also adds the necessary plumbing for asynchronous conversions. Plugins can optionally implement the `convert_async` function which calls a callback when the conversion is complete.

The new `NoiseTextureConversionPlugin` implements this. It opens a dialog to allow the user to select the save location of the new file.

### Alternative implementations
I considered 3 possible implementations of async conversions:
* Add a new `EditorResourceAsyncConversionPlugin` in addition to the existing `EditorResourceConversionPlugin`. Because the interface is different it would require doubling up all the places conversion plugins are currently plumbed in (the filesystem dock, the resource picker, the `find_resource_conversion_plugin` functions etc) with an async version of each.
* Modify the existing `EditorResourceConversionPlugin` to always support asynchronous conversions. `EditorResourceConversionPlugin::convert` would always take a callback, and the existing non-async plugins could just call it immediately.
* Add `convert_async` to the existing `EditorResourceConversionPlugin`, which may return null when the plugin doesn't support async conversions. Likewise, plugins could return null for `convert` if they don't support normal conversions. This way users can try an async conversion if a normal conversion fails.

I opted for the third option because it involved the smallest change to existing code. Option one would involve a lot of almost-duplicate code which didn't seem very elegant. Option two is probably the cleanest but would be an API break and involve changing all existing conversion plugins - I'm a first-time contributor so I don't know what the appetite is for this kind of change

I am more than happy to try one of the other proposed implementations (or indeed anything else) if that would be more desirable.

callumfare committed:
Implement baking of noise textures

Calinou commented:
- Related to https://github.com/godotengine/godot-proposals/issues/7023.

Note that unlike https://github.com/godotengine/godot-proposals/issues/7023, this is not a transparent approach to caching. While it avoids incurring a loading time penalty for cold loads (with no cache), it does mean you'll have to commit a large binary file to version control, which bloats VCS history.

cfare commented:
@Calinou Thanks, I had a look for existing proposals but somehow didn't find that one, that's helpful context! 

In my use cases the size of the generated file didn't really come up as an issue as it's just one extra resource among many others. I can see how caching would be preferable if that is a problem though.

To address the destructive nature I could add an option in the dialog to preserve the original resource, even though all its uses will be replaced. This would be a difference in behavior from the other plugins though which is maybe not desirable. 

Switching to WebP seems like it makes sense based on what you said, I'll look into that.

I also see I've missed some documentation, I'll address that too.

beicause commented:
I think this PR is great. At least for those who need it, they don't have to use scripts for conversion.
 
I found a issue that conversion will crash if NoiseTexture is pink placeholder (noise is null).

This can implements conversion for GradientTexture. In addition, for usability, when the parent directory doesn't exist, it should automatically create it. And when the target path already exists, it should allow to overwrite it.

callumfare committed:
Address review feedback

cfare commented:
Apologies for taking a while to come back to this!

@beicause Thanks for the feedback. I've added a check for if the image is empty, and added an option to allow overwriting existing files.

@Calinou I've switched the default format to WebP, but PNG is still a supported option just in case. In my opinion this adds a workflow that will be helpful for some users despite the drawbacks. It's simple to understand what it does, so the destructive nature won't be a surprise to users. Realistically I don't have the time or familiarity with Godot to try implementing the caching proposal, so I'll leave it up to you and the other maintainers to decide if this is worth merging. If you do I'm still happy to change to the alternative implementations I mentioned in the PR description!

callumfare committed:
Fix style errors

beicause commented:
It is worth mentioning that there is a better and simple way to automatically bake all the noise textures during export and remain editable in editor. https://github.com/godotengine/godot-proposals/issues/7023#issuecomment-2817433102
